### DISCUSSION TITLE
✨ CI: dogfooding feedback

### DISCUSSION THREAD

**Discussion by aluzzardi** [dagger/dagger#2080] [General]
Dogfooding notes/wishlist from #1984

/cc @shykes @helderco @samalba

- I wish I could do `dagger do build lint` (a la `Makefile`)
- I wish console outputs were available. It would have been a tremendous help to debug. Also very helpful to inform what just happened (e.g. `dagger do build` should print the revision we just built). I hacked around by writing a file and then removed that line before committing 
- I wish the platform defaulted to buildkitd's platform. The performance is horrible in any other setting. I know we had long discussions around the API for this, but it's so important my recommendation would be to default to buildkitd and provide a `--platform` flag (so we don't need to define the whole CUE DX before making this change)
- It would be nice if dagger was GitHub-aware. I split `build` and `lint` into two different GH jobs because I want to see them separately in the status. However, that doesn't do any good with caching. Ideally the dagger GHA could report individual actions as "statuses" in the PR?
- The `docker.#Run` / `docker.#Build` experience was a bit challenging: not sure when to use one or the other. On the one hand, I prefer using`dagger.#Build` to define pipelines (e.g. do this, do that, XXX) -- much better than `_doA: ... _doB: ...`. On the other hand, I can't do that as soon as I want to export files. I started writing `docker.#Build` steps then had to convert to manual operations (see `actions.version`)
- Also, I'm using `docker.#Build` like a `dagger.#Pipeline` -- I'm not really building anything
- Converting this line from our Makefile resulted in 40 lines of CUE: `go build -ldflags '-s -w -X go.dagger.io/dagger/version.Revision=$(git rev-parse --short HEAD)`. Mostly because of the point above
- Language Server: I spent my time jumping between upstream definitions to figure out the interface. I was able to do that because universe happens to be in the same repo as the CI config, anything else would have been a major PITA. Having auto-completion would have been a life saver.

**Comment by helderco** [comment]
Great stuff 🙂 

> Dogfooding notes/wishlist @shykes @helderco @samalba:
> 
>    * I wish I could do `dagger do build lint` (a la `Makefile`)

Me too! And it makes it more portable. ([windows anyone](https://github.com/dagger/dagger/pull/1783#issuecomment-1072716180)?)

>    * I wish console outputs were available. It would have been a tremendous help to debug. Also very helpful to inform what just happened (e.g. `dagger do build` should print the revision we just built). I hacked around by writing a file and then removed that line before committing

Yeah, that's also what I've been doing. You know I'm partial to this 🙂 

>    * I wish the platform defaulted to buildkitd's platform. The performance is horrible in any other setting. I know we had long discussions around the API for this, but it's so important my recommendation would be to default to buildkitd and provide a `--platform` flag (so we don't need to define the whole CUE DX before making this change)

Not aware of this. Want to know more.

>    * It would be nice if dagger was GitHub-aware. I split `build` and `lint` into two different GH jobs because I want to see them separately in the status. However, that doesn't do any good with caching. Ideally the dagger GHA could report individual actions as "statuses" in the PR?

This has been mentioned by multiple users.

>    * The `docker.#Run` / `docker.#Build` experience was a bit challenging: not sure when to use one or the other. On the one hand, I prefer using`dagger.#Build` to define pipelines (e.g. do this, do that, XXX) -- much better than `_doA: ... _doB: ...`. On the other hand, I can't do that as soon as I want to export files. I started writing `docker.#Build` steps then had to convert to manual operations (see `actions.version`)

Interesting. I've been through this in January a lot, but it was because of the nesting issue. I know how frustrating it is to break those steps into correctly hooked fields. One small typo and you have an error that's not the root cause (incorrect `#FS` leading to errors elsewhere).

>    * Also, I'm using `docker.#Build` like a `dagger.#Pipeline` -- I'm not really building anything

Yes, if there was a generic thing to hook any `input: _` it could be reused for other things with the same workflow. I've done this in an abstraction for adding files to an `#FS` (instead of `#Image`).

>    * Converting this line from our Makefile resulted in 40 lines of CUE: `go build -ldflags '-s -w -X go.dagger.io/dagger/version.Revision=$(git rev-parse --short HEAD)`. Mostly because of the point above

I'll look into that more closely.

>    * Language Server: I spent my time jumping between upstream definitions to figure out the interface. I was able to do that because universe happens to be in the same repo as the CI config, anything else would have been a major PITA. Having auto-completion would have been a life saver.

Yes, and having full reference docs could help as well.

**Comment by aluzzardi** [comment]
> I wish the platform defaulted to buildkitd's platform. The performance is horrible in any other setting. I know we had long discussions around the API for this, but it's so important my recommendation would be to default to buildkitd and provide a --platform flag (so we don't need to define the whole CUE DX before making this change)

>> Not aware of this. Want to know more.

Opened a separate issue #1996 with more details

**Comment by jpluscplusm** [comment]
Console outputs (or another hermetic-seal-breaking-but-not-really equivalent): **very** hard agree!

I'm trying to debug a universe/aws.cli run that's working locally but failing in GHA, and I'm almost out of ideas for how to surface the underlying AWS response error.

As a slight extra ask in the console output space: it'd be great to have a way to unconditionally print some output from a step, regardless of if the DAG failed at that step. My current strategy breaks (I think!) because the `client: filesystem: foo: write: contents: actions.failingstep.output` node never gets reached, leaving me fumbling around blind right now!

**Comment by helderco** [comment]
> Console outputs (or another hermetic-seal-breaking-but-not-really equivalent): **very** hard agree!

Context:
- #1351
- https://github.com/dagger/dagger/pull/1486#issuecomment-1081551488

**Comment by helderco** [comment]
> - Converting this line from our Makefile resulted in 40 lines of CUE: `go build -ldflags '-s -w -X go.dagger.io/dagger/version.Revision=$(git rev-parse --short HEAD)`. Mostly because of the point above

You could also use the local host:

```cue
client: commands: gitVersion: {
	name: "git"
	args: ["rev-parse", "--short", "HEAD"]
}

actions: {
	build: go.#Build & {
		ldflags: "-s -w -X go.dagger.io/dagger/version.Revision=\(client.commands.gitVersion.stdout)"
	}
}
```

But this would require git on the host, and I get that it's better to have as much as possible running in a container. To that end, we could have a convenience in `go.#Build` to extract the version like that, but I think it's better for companies to make their own conventions in their own high-level definitions.

```cue
#GoBuild & {
    source: dagger.#FS
    _version: ...
    _build: go.#Build & {
        "source":  source
        ldflags: ...
    }
}
```

In our case, if it's done only once, I don't feel there's a need to create an abstraction. We do have plans to make fetching the version slightly better (no need for bash):

```cue
_image: go.#Image // or other image with git

version: docker.#Run & {
	input:   _image.output
        // still need to mount the src
	workdir: "/src"
	mounts: source: {
		dest:     "/src"
		contents: _source
	}
	command: {
		name: "git"
		args: ["rev-parse", "--short", "HEAD"]
	}
	// less boilerplate to get output (#1620)
	stdout: string
}

build: go.#Build & {
	source:  _source
	ldflags: "-s -w -X go.dagger.io/dagger/version.Revision=\(version.stdout)"
}
```

Maybe a `git.#Run` for running git commands against some source code (from `universe.dagger.io/git`):

```cue
version: git.#Run & {
	source:  _source
	command: args: ["rev-parse", "--short", "HEAD"]
	stdout: string
}

build: go.#Build & {
	source:  _source
	ldflags: "-s -w -X go.dagger.io/dagger/version.Revision=\(version.stdout)"
}
```
